File: TeamCode/src/main/java/barrel/enumerators/Ball.java
package barrel.enumerators;


public class Ball
{
    static public final int NONE = 0, PURPLE = 1, GREEN = 2;

    private Name _memName;
    private int  _memId;


    public Ball(int value)
    {
        Set(ToName(value), value);
    }
    public Ball(Name name)
    {
        Set(name, ToInt(name));
    }
    public Ball(int value, Name name)
    {
        Set(name, value);
    }
    public Ball(Name name, int value)
    {
        Set(name, value);
    }
    public Ball()
    {
        Empty();
    }
    public enum Name
    {
        NONE,
        PURPLE,
        GREEN
    }



    public void Empty()
    {
        Set(Name.NONE, NONE);
    }

    public void Set(int value)
    {
        Set(ToName(value), value);
    }
    public void Set(Name name)
    {
        Set(name, ToInt(name));
    }
    public void Set(int value, Name name)
    {
        Set(name, value);
    }
    public void Set(Name name, int value)
    {
        _memId = value;
        _memName = name;
    }


    public Name Name()
    {
        return _memName;
    }
    public int  Id()
    {
        return _memId;
    }

    public boolean IsFilled()
    {
        return _memId != NONE;
    }
    public boolean HasBall()
    {
        return  _memId != NONE;
    }
    public boolean HasBall(Ball ball)
    {
        return _memId == ball.Id();
    }
    public boolean HasBall(Ball.Name ball)
    {
        return _memName == ball;
    }
    public boolean HasPurpleBall()
    {
        return _memId == PURPLE;
    }
    public boolean HasGreenBall()
    {
        return  _memId == GREEN;
    }
    public boolean IsEmpty()
    {
        return _memId == NONE;
    }


    static public Name ToName (int value)
    {
        return value == NONE ? Name.NONE
                : value == PURPLE ? Name.PURPLE
                : Name.GREEN;
    }
    static public int ToInt (Name name)
    {
        switch (name)
        {
            case NONE:   return NONE;
            case PURPLE: return PURPLE;
            default:     return GREEN;
        }
    }
}

File: TeamCode/src/main/kotlin/org/woen/enumerators/Ball.kt
package org.woen.enumerators



enum class BallRequest
{
    PURPLE,
    GREEN,

    PREFER_PURPLE,
    PREFER_GREEN,
    ANY_CLOSEST,

    NONE
}


class Ball
{
    private var _id   = NONE
    private var _name = Name.NONE


    constructor() { empty() }
    constructor(id:   Int,  name: Name = toName(id))  { set(id, name) }
    constructor(name: Name, id:   Int  = toInt(name)) { set(id, name) }


    enum class Name
    {
        PURPLE,
        GREEN,

        UNKNOWN_COLOR,
        NONE
    }


    fun empty() = set(NONE, Name.NONE)

    fun set(name: Name,  id: Int  = toInt(name)) = set(id, name)
    fun set(id:   Int, name: Name = toName(id))
    {
        _id   = id
        _name = name
    }
    fun set(ball: Ball)
    {
        _id   = ball.id
        _name = ball.name
    }


    val id   get() = _id
    val name get() = _name

    fun formattedName(): String
    {
        return when (_name)
        {
            Name.PURPLE        -> "+ PURPLE +"
            Name.GREEN         -> "+ GREEN + "

            Name.UNKNOWN_COLOR -> "  + ?? +  "
            Name.NONE          -> "    --    "
        }
    }


    fun isFilled() = _id != NONE
    fun isEmpty()  = _id == NONE


    fun hasBall(name: Name) = _name == name


    fun isPseudoMatch(name: BallRequest)
        = !isEmpty() &&
            (
                isAbstractAny(name) ||
                hasBall(toBall(name))
            )
    fun isTrueMatch(name: BallRequest)
        = !isEmpty() && hasBall(toBall(name))



    companion object
    {
        const val PURPLE: Int = 0
        const val GREEN:  Int = 1

        const val UNKNOWN_COLOR: Int = 2
        const val NONE:          Int = 3


        fun toName(id:  Int): Name
            = when (id)
            {
                PURPLE -> Name.PURPLE
                GREEN  -> Name.GREEN

                UNKNOWN_COLOR -> Name.UNKNOWN_COLOR
                else          -> Name.NONE
            }
        fun toInt(name: Name): Int
            = when (name)
            {
                Name.PURPLE -> PURPLE
                Name.GREEN  -> GREEN

                Name.UNKNOWN_COLOR -> UNKNOWN_COLOR
                Name.NONE          -> NONE
            }
        fun toBall(name: BallRequest): Name
            = when (name)
                {
                    BallRequest.PURPLE, BallRequest.PREFER_PURPLE -> Name.PURPLE
                    BallRequest.GREEN,  BallRequest.PREFER_GREEN  -> Name.GREEN

                    BallRequest.ANY_CLOSEST -> Name.UNKNOWN_COLOR
                    BallRequest.NONE        -> Name.NONE
                }

        fun isAbstractAny(name: BallRequest)
            =   name == BallRequest.PREFER_PURPLE ||
                name == BallRequest.PREFER_GREEN  ||
                name == BallRequest.ANY_CLOSEST
    }
}